File: src/lib/sequential/SkipList.ml
type 'a node_val =
  | Start
  | End
  | Val of 'a

let compare_node_val a b =
  match (a, b) with
  | Start, Start -> 0
  | Start, _ -> -1
  | _, Start -> 1
  | End, End -> 0
  | End, _ -> 1
  | _, End -> -1
  | Val x, Val y -> Stdlib.compare x y

type 'a node = {
  value : 'a node_val;
  forward : 'a node option array;
}

type 'a skip_list = {
  mutable level : int;
  max_level : int;
  prob : float;
  head : 'a node;
  tail : 'a node;
}

let empty max_level prob =
  if max_level < 1 then invalid_arg "SkipList.empty: max_level must be >= 1";
  if prob <= 0.0 || prob >= 1.0 then
    invalid_arg "SkipList.empty: prob must be in (0.0, 1.0)";
  Random.self_init ();
  let tail = { value = End; forward = Array.make (max_level + 1) None } in
  let head = { value = Start; forward = Array.make (max_level + 1) (Some tail) } in
  { level = 0; max_level; prob; head; tail }

let create max_level prob = empty max_level prob

let random_level sl =
  let rec aux lvl =
    if lvl < sl.max_level && Random.float 1.0 < sl.prob then aux (lvl + 1)
    else lvl
  in
  aux 0

let search sl target =
  let curr = ref sl.head in
  for i = sl.level downto 0 do
    let continue = ref true in
    while !continue do
      match (!curr).forward.(i) with
      | Some node when compare_node_val node.value (Val target) < 0 -> curr := node
      | _ -> continue := false
    done
  done;
  match (!curr).forward.(0) with
  | Some node when compare_node_val node.value (Val target) = 0 -> true
  | _ -> false

let insert sl value =
  let update = Array.make (sl.max_level + 1) sl.head in
  let curr = ref sl.head in

  for i = sl.level downto 0 do
    let continue = ref true in
    while !continue do
      match (!curr).forward.(i) with
      | Some node when compare_node_val node.value (Val value) < 0 -> curr := node
      | _ -> continue := false
    done;
    update.(i) <- !curr
  done;

  let next0 = (!curr).forward.(0) in
  match next0 with
  | Some node when compare_node_val node.value (Val value) = 0 -> ()
  | _ ->
    let new_level = random_level sl in

    if new_level > sl.level then begin
      for i = sl.level + 1 to new_level do
        update.(i) <- sl.head
      done;
      sl.level <- new_level
    end;

    let new_node =
      { value = Val value; forward = Array.make (new_level + 1) None }
    in

    for i = 0 to new_level do
      new_node.forward.(i) <- update.(i).forward.(i);
      update.(i).forward.(i) <- Some new_node
    done

let erase sl value =
  let update = Array.make (sl.max_level + 1) sl.head in
  let curr = ref sl.head in

  for i = sl.level downto 0 do
    let continue = ref true in
    while !continue do
      match (!curr).forward.(i) with
      | Some node when compare_node_val node.value (Val value) < 0 -> curr := node
      | _ -> continue := false
    done;
    update.(i) <- !curr
  done;

  match (!curr).forward.(0) with
  | Some node when compare_node_val node.value (Val value) = 0 ->
    for i = 0 to sl.level do
      match update.(i).forward.(i) with
      | Some n when n == node -> update.(i).forward.(i) <- node.forward.(i)
      | _ -> ()
    done;

    while sl.level > 0 && update.(0) == sl.head && sl.head.forward.(sl.level) == Some sl.tail do
      sl.level <- sl.level - 1
    done
  | _ -> ()

let to_lists sl =
  let collect_level i =
    let rec loop acc node_opt =
      match node_opt with
      | None -> List.rev acc
      | Some node -> (
        match node.value with
        | End -> List.rev acc
        | Start -> loop acc node.forward.(i)
        | Val v -> loop (v :: acc) node.forward.(i)
      )
    in
    loop [] sl.head.forward.(i)
  in
  let rec all_levels i acc =
    if i < 0 then acc else all_levels (i - 1) ((i, collect_level i) :: acc)
  in
  all_levels sl.level []

let print pp_value sl =
  for i = sl.level downto 0 do
    Printf.printf "Level %d: " i;
    let rec loop node_opt =
      match node_opt with
      | None -> ()
      | Some node -> (
        match node.value with
        | End -> ()
        | Start -> loop node.forward.(i)
        | Val v ->
          pp_value v;
          Printf.printf " ";
          loop node.forward.(i)
      )
    in
    loop sl.head.forward.(i);
    Printf.printf "\n"
  done

File: src/lib/sequential/skiplist.cpp
#include <bits/stdc++.h>
using namespace std;

class SkipList {
private:

    /* 
        Node structure:
        Each node stores:
        - value
        - forward pointers (one for each level)
    */
    struct Node {
        int value;
        vector<Node*> forward;

        // Constructor
        Node(int val, int level) {
            value = val;
            forward.resize(level + 1, nullptr);
        }
    };

    int MAX_LEVEL;     // Maximum level allowed in skiplist
    float P;           // Probability for random level generation
    int level;         // Current highest level in skiplist
    Node* header;      // Header node (start of skiplist)

    /*
        Generate random level for new node
        With probability P, increase level
    */
    int randomLevel() {
        int lvl = 0;

        // Keep increasing level with probability P
        while (((float)rand() / RAND_MAX) < P && lvl < MAX_LEVEL)
            lvl++;

        return lvl;
    }

public:

    /*
        Constructor
        Initialize skiplist with:
        - max level
        - probability
    */
    SkipList(int maxLevel = 16, float prob = 0.5) {
        MAX_LEVEL = maxLevel;
        P = prob;
        level = 0;

        // Create header node with max level
        header = new Node(-1, MAX_LEVEL);
    }


    /*
        SEARCH operation
        Start from top level and move down
    */
    bool search(int target) {

        Node* curr = header;

        // Traverse from highest level to lowest
        for (int i = level; i >= 0; i--) {

            // Move forward while value is smaller
            while (curr->forward[i] != nullptr &&
                   curr->forward[i]->value < target) {

                curr = curr->forward[i];
            }
        }

        // Move to next node at level 0
        curr = curr->forward[0];

        // Check if found
        if (curr != nullptr && curr->value == target)
            return true;

        return false;
    }


    /*
        INSERT operation
    */
    void insert(int value) {

        // Stores nodes that need pointer updates
        vector<Node*> update(MAX_LEVEL + 1);

        Node* curr = header;

        // Find positions where node should be inserted
        for (int i = level; i >= 0; i--) {

            while (curr->forward[i] != nullptr &&
                   curr->forward[i]->value < value) {

                curr = curr->forward[i];
            }

            update[i] = curr;
        }

        // Move to level 0
        curr = curr->forward[0];

        // Insert only if value not already present
        if (curr == nullptr || curr->value != value) {

            // Generate random level
            int newLevel = randomLevel();

            // If new level higher, initialize header pointers
            if (newLevel > level) {
                for (int i = level + 1; i <= newLevel; i++)
                    update[i] = header;

                level = newLevel;
            }

            // Create new node
            Node* newNode = new Node(value, newLevel);

            // Update forward pointers
            for (int i = 0; i <= newLevel; i++) {

                newNode->forward[i] = update[i]->forward[i];
                update[i]->forward[i] = newNode;
            }
        }
    }


    /*
        DELETE operation
    */
    void erase(int value) {

        vector<Node*> update(MAX_LEVEL + 1);
        Node* curr = header;

        // Find node to delete
        for (int i = level; i >= 0; i--) {

            while (curr->forward[i] != nullptr &&
                   curr->forward[i]->value < value) {

                curr = curr->forward[i];
            }

            update[i] = curr;
        }

        curr = curr->forward[0];

        // If node found
        if (curr != nullptr && curr->value == value) {

            // Update forward pointers
            for (int i = 0; i <= level; i++) {

                // If pointer doesn't match, stop
                if (update[i]->forward[i] != curr)
                    break;

                update[i]->forward[i] = curr->forward[i];
            }

            // Delete node
            delete curr;

            // Adjust level if needed
            while (level > 0 &&
                   header->forward[level] == nullptr) {

                level--;
            }
        }
    }


    /*
        Print skiplist (for debugging)
    */
    void print() {
        cout << "\nSkip List:\n";

        for (int i = level; i >= 0; i--) {

            Node* node = header->forward[i];

            cout << "Level " << i << ": ";

            while (node != nullptr) {
                cout << node->value << " ";
                node = node->forward[i];
            }

            cout << endl;
        }
    }
};



/*
    Driver Code
*/
int main() {

    SkipList sl;

    // Insert elements
    sl.insert(3);
    sl.insert(6);
    sl.insert(7);
    sl.insert(9);
    sl.insert(12);
    sl.insert(19);

    sl.print();

    // Search
    cout << "\nSearch 6: " << sl.search(6) << endl;
    cout << "Search 15: " << sl.search(15) << endl;

    // Delete
    sl.erase(6);

    cout << "\nAfter deleting 6:";
    sl.print();

    return 0;
}